demonstrations/tutorial_qjit_compile_grovers_algorithm_with_catalyst.py

```diff
@@ -0,0 +1,340 @@
+# containing all possible :math:`n`-qubit states.
+#
+# First, we'll import the required packages and define the Grover's algorithm circuit, as we did in
+# the previous tutorial:
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pennylane as qml
+
+
+def equal_superposition(wires):
+    for wire in wires:
+        qml.Hadamard(wires=wire)
+
+
+def oracle(wires, omega):
+    qml.FlipSign(omega, wires=wires)
+
+
+def num_grover_iterations(N, M):
+    return np.ceil(np.sqrt(N / M) * np.pi / 4).astype(int)
+
+
+def grover_circuit(num_qubits):
+    wires = list(range(num_qubits))
+    omega = np.array([np.zeros(num_qubits), np.ones(num_qubits)])
+
+    M = len(omega)
+    N = 2**num_qubits
+
+    # Initial state preparation
+    equal_superposition(wires)
+
+    # Grover iterations
+    for _ in range(num_grover_iterations(N, M)):
+        for omg in omega:
+            oracle(wires, omg)
+        qml.templates.GroverOperator(wires)
+
+    return qml.probs(wires=wires)
+
+
+######################################################################
+# We'll begin with a circuit defined using the default state-simulator device, ``"default.qubit"``,
+# as our baseline. See the documentation in :func:`~.pennylane.device` for a list of other supported
+# devices. To run our performance benchmarks, we'll increase the number of qubits in our circuit to
+# :math:`n = 12`.
+
+NUM_QUBITS = 12
+
+
+@qml.qnode(qml.device("default.qubit", wires=NUM_QUBITS))
+def circuit_default_qubit():
+    return grover_circuit(NUM_QUBITS)
+
+
+results = circuit_default_qubit()
+
+
+######################################################################
+# Let's quickly confirm that Grover's algorithm correctly identified the solution states
+# :math:`\vert 0\rangle ^{\otimes n}` and :math:`\vert 1\rangle ^{\otimes n}` as the most likely
+# states to be measured:
+
+most_probable_states = np.argsort(results)[-2:][::-1]
+
+for i in most_probable_states:
+    print(f"Prob of state '{i:0{NUM_QUBITS}b}': {results[i]:.4g}")
+
+
+######################################################################
+# It worked! We are now ready to QJIT compile our Grover's algorithm circuit.
+
+
+######################################################################
+# Quantum Just-In-Time Compiling the Circuit
+# ------------------------------------------
+#
+# At the time of writing, Catalyst does not support the ``"default.qubit"`` state-simulator device,
+# so let's first define a new circuit using `Lightning
+# <https://docs.pennylane.ai/projects/lightning/en/stable/index.html>`__, which is a PennyLane
+# plugin that provides more performant state simulators written in C++, and which is supported by
+# Catalyst. The full list of devices supported by Catalyst is available `here
+# <https://docs.pennylane.ai/projects/catalyst/en/stable/dev/devices.html>`__.
+
+
+@qml.qnode(qml.device("lightning.qubit", wires=NUM_QUBITS))
+def circuit_lightning():
+    return grover_circuit(NUM_QUBITS)
+
```

Comment:
```suggestion
dev = qml.device("lightning.qubit", wires=NUM_QUBITS)

@qml.qnode(dev)
def circuit_lightning():
    return grover_circuit(NUM_QUBITS)
```

Reply:
Similarly, you could also consider

```suggestion
dev = qml.device("lightning.qubit", wires=NUM_QUBITS)
circuit_lightning = qml.QNode(dev, lambda: grover_circuit(NUM_QUBITS))
```